File: MoviesExample/core/src/main/java/com/novoda/demo/movies/MovieService.java
package com.novoda.demo.movies;

import com.novoda.demo.movies.api.MoviesApi;
import com.novoda.demo.movies.api.MoviesResponse;
import com.novoda.demo.movies.api.VideosResponse;
import com.novoda.demo.movies.model.Movie;
import com.novoda.demo.movies.model.Video;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Response;
import rx.Observable;
import rx.Observer;
import rx.functions.Func1;
import rx.schedulers.Schedulers;

public class MovieService {

    private final MoviesApi api;
    private MoviesSate moviesSate = new MoviesSate(new ArrayList<Movie>(), 1);

    private Callback callback;

    public MovieService(MoviesApi api) {
        this.api = api;
    }

    public void subscribe(Callback callback) {
        this.callback = callback;
        callback.onNewData(moviesSate);
        if (moviesSate.isEmpty()) {
            loadMore();
        }
    }

    public void unsubscribe(Callback callback) {
        this.callback = null;
    }

    public void loadMore() {
        api.topRated(moviesSate.pageNumber()).enqueue(new retrofit2.Callback<MoviesResponse>() {
            @Override
            public void onResponse(Call<MoviesResponse> call, Response<MoviesResponse> response) {
                if (response == null || response.body() == null || response.body().results == null) {
                    return;
                }
                List<Movie> movies = moviesSate.movies();
                movies.addAll(response.body().results);
                moviesSate = new MoviesSate(movies, moviesSate.pageNumber() + 1);
                callback.onNewData(moviesSate);
            }

            @Override
            public void onFailure(Call<MoviesResponse> call, Throwable throwable) {
                callback.onFailure(throwable);
            }
        });
    }

    public void loadTrailerFor(Movie movie, final TrailerCallback trailerCallback) {
        api.videos(movie.id)
                .flatMapObservable(new Func1<VideosResponse, Observable<Video>>() {
                    @Override
                    public Observable<Video> call(final VideosResponse videosResponse) {
                        return Observable.from(videosResponse.results);
                    }
                })
                .takeFirst(new Func1<Video, Boolean>() {
                    @Override
                    public Boolean call(final Video video) {
                        return video.trailerUrl() != null;
                    }
                })
                .subscribeOn(Schedulers.io())
                .subscribe(new Observer<Video>() {
                    @Override
                    public void onCompleted() {
                        // noop
                    }

                    @Override
                    public void onError(final Throwable e) {
                        trailerCallback.onFailure(e);
                    }

                    @Override
                    public void onNext(final Video video) {
                        trailerCallback.onTrailerLoaded(video);
                    }
                });
         }

    interface Callback {
        void onNewData(MoviesSate moviesSate);

        void onFailure(Throwable e);
    }

    interface TrailerCallback {
        void onTrailerLoaded(Video video);

        void onFailure(Throwable e);
    }

}

File: MoviesExample/core/src/main/java/com/novoda/demo/movies/MovieService.kt
package com.novoda.demo.movies

import com.novoda.demo.movies.api.MoviesApi
import com.novoda.demo.movies.api.MoviesResponse
import com.novoda.demo.movies.model.Movie
import com.novoda.demo.movies.model.Video
import retrofit2.Call
import retrofit2.Response
import rx.Observable
import rx.Observer
import rx.schedulers.Schedulers
import java.util.ArrayList

class MovieService(private val api: MoviesApi) {
    private var moviesState = MoviesState(ArrayList(), 1)

    private var callback: Callback? = null

    fun subscribe(callback: Callback) {
        this.callback = callback
        callback.onNewData(moviesState)
        if (moviesState.isEmpty) {
            loadMore()
        }
    }

    fun unsubscribe(callback: Callback) {
        this.callback = null
    }

    fun loadMore() {
        api.topRated(moviesState.pageNumber()).enqueue(object : retrofit2.Callback<MoviesResponse> {
            override fun onResponse(call: Call<MoviesResponse>, response: Response<MoviesResponse>?) {
                if (response?.body() == null || response.body()?.results == null) {
                    return
                }
                val movies = moviesState.movies()
                movies.addAll(response.body()!!.results)
                moviesState = MoviesState(movies, moviesState.pageNumber() + 1)
                callback!!.onNewData(moviesState)
            }

            override fun onFailure(call: Call<MoviesResponse>, throwable: Throwable) {
                callback!!.onFailure(throwable)
            }
        })
    }

    fun loadTrailerFor(movie: Movie, trailerCallback: TrailerCallback) {
        api.videos(movie.id)
                .flatMapObservable { videosResponse -> Observable.from(videosResponse.results) }
                .takeFirst { video -> video.trailerUrl() != null }
                .subscribeOn(Schedulers.io())
                .subscribe(object : Observer<Video> {
                    override fun onCompleted() {
                        // noop
                    }

                    override fun onError(e: Throwable) {
                        trailerCallback.onFailure(e)
                    }

                    override fun onNext(video: Video) {
                        trailerCallback.onTrailerLoaded(video)
                    }
                })
    }

    interface Callback {
        fun onNewData(moviesSate: MoviesState)

        fun onFailure(e: Throwable)
    }

    interface TrailerCallback {
        fun onTrailerLoaded(video: Video)

        fun onFailure(e: Throwable)
    }

}
